In [7]:
#Importando pandas e importando tabela
import pandas as pd 
tabela = pd.read_csv("cancelamentos_sample.csv")
tabela = tabela.drop("CustomerID",axis=1)
display(tabela)

idade    sexo  tempo_como_cliente  frequencia_uso  ligacoes_callcenter  \
0       23.0    Male                13.0            22.0                  2.0   
1       49.0    Male                55.0            16.0                  3.0   
2       30.0    Male                 7.0             1.0                  0.0   
3       26.0    Male                40.0             5.0                  3.0   
4       27.0  Female                17.0            30.0                  5.0   
...      ...     ...                 ...             ...                  ...   
49995   62.0  Female                35.0             7.0                  2.0   
49996   36.0    Male                43.0            21.0                  2.0   
49997   55.0    Male                42.0             8.0                  1.0   
49998   40.0  Female                14.0            19.0                  1.0   
49999   64.0    Male                41.0            29.0                  5.0   

       dias_atraso assinatura duracao_contrato  total_gasto  \
0              1.0   Standard           Annual       909.58   
1              6.0    Premium          Monthly       207.00   
2              8.0      Basic           Annual       768.78   
3              8.0    Premium           Annual       398.00   
4              6.0      Basic           Annual       507.00   
...            ...        ...              ...          ...   
49995          8.0      Basic           Annual       232.00   
49996         30.0      Basic        Quarterly       928.00   
49997         12.0      Basic          Monthly       326.00   
49998         17.0    Premium        Quarterly       826.76   
49999          9.0      Basic        Quarterly       700.38   

       meses_ultima_interacao  cancelou  
0                        23.0       0.0  
1                        29.0       1.0  
2                         7.0       0.0  
3                        12.0       1.0  
4                        15.0       1.0  
...                       ...       ...  
49995                    15.0       1.0  
49996                    30.0       1.0  
49997                    27.0       1.0  
49998                    12.0       0.0  
49999                    18.0       1.0  

[50000 rows x 11 columns]

In [ ]:
#Removendo Valores Vazios (NaN)
display(tabela.info())
tabela = tabela.dropna()
display(tabela.info())

In [ ]:
# Verificando a Coluna de Cancelamentos
display(tabela["cancelou"].value_counts())
display(tabela["cancelou"].value_counts(normalize=True).map("{:.1%}".format))

In [ ]:
#Verificando a Quantidade da Coluna Duração_contrato
display(tabela["duracao_contrato"].value_counts(normalize=True))
display(tabela["duracao_contrato"].value_counts())

In [ ]:
#Analisando as informações dos contratos
display(tabela.groupby("duracao_contrato").mean(numeric_only=True))
#a media de Cancelamentos e igual a 1, ou seja Todos os contratos mensais forma cancelados

Verificando a coluna contratos sem o contrato mensal,Sabendo que o contrato mensal é ruim para a empresa, 
nós podemos remover as informações desse contrato específico e continuar analisando.


In [ ]:

tabela= tabela[tabela["duracao_contrato"]!="Monthly"]
display(tabela)
display(tabela["cancelou"].value_counts())
display(tabela["cancelou"].value_counts(normalize=True).map("{:.1%}".format))

Como ainda temos um número bem alto de cancelamentos vamos agora fazer uma análise nas assinaturas para verificar se podemos tirar 
alguma conclusão para melhorar esse índice de cancelamentos.

In [ ]:
# Utilização graficos Para A Analise, Ira Criar um grafico para cada coluna(11 graficos) 
import plotly.express as px

for coluna in tabela.columns:
    grafico = px.histogram(tabela, x=coluna, color="cancelou",width=600)
    grafico.show()
    
#Com os Graficos e possivel observar que clientes com mais de 20 dias de atraso cancelam suas assinaturas. 
# E clientes com mais de 5 ligações ao call center cancelam suas assinaturas.      


Analisando apenas os dois gráficosindicados, nós podemos manter na tabela as ligações ao call center que são menoresdo que 5.
E podemos manter também asinformações onde os dias de atraso sãomenores do que 0.
Com isso podemos visualizar a base de dados e fazer o cálculo novamente para verificar o percentual de cancelamento.

In [8]:
# com os graficos a gente consegue descobrir muita coisa:
# dias atraso acima de 20 dias, 100% cancela
# ligações call center acima de 5 todo mundo cancela

tabela = tabela[tabela["ligacoes_callcenter"]<5]
tabela = tabela[tabela["dias_atraso"]<20]
display(tabela)
display(tabela["cancelou"].value_counts())
display(tabela["cancelou"].value_counts(normalize=True).map("{:.1%}".format))


idade    sexo  tempo_como_cliente  frequencia_uso  ligacoes_callcenter  \
0       23.0    Male                13.0            22.0                  2.0   
1       49.0    Male                55.0            16.0                  3.0   
2       30.0    Male                 7.0             1.0                  0.0   
3       26.0    Male                40.0             5.0                  3.0   
5       30.0     NaN                 NaN            12.0                  3.0   
...      ...     ...                 ...             ...                  ...   
49992   37.0    Male                46.0            29.0                  2.0   
49994   63.0    Male                16.0            24.0                  2.0   
49995   62.0  Female                35.0             7.0                  2.0   
49997   55.0    Male                42.0             8.0                  1.0   
49998   40.0  Female                14.0            19.0                  1.0   

       dias_atraso assinatura duracao_contrato  total_gasto  \
0              1.0   Standard           Annual       909.58   
1              6.0    Premium          Monthly       207.00   
2              8.0      Basic           Annual       768.78   
3              8.0    Premium           Annual       398.00   
5              7.0   Standard        Quarterly       641.78   
...            ...        ...              ...          ...   
49992          9.0   Standard        Quarterly       517.86   
49994         18.0   Standard        Quarterly       442.00   
49995          8.0      Basic           Annual       232.00   
49997         12.0      Basic          Monthly       326.00   
49998         17.0    Premium        Quarterly       826.76   

       meses_ultima_interacao  cancelou  
0                        23.0       0.0  
1                        29.0       1.0  
2                         7.0       0.0  
3                        12.0       1.0  
5                         7.0       0.0  
...                       ...       ...  
49992                     3.0       0.0  
49994                    26.0       1.0  
49995                    15.0       1.0  
49997                    27.0       1.0  
49998                    12.0       0.0  

[27916 rows x 11 columns]

cancelou
0.0    20392
1.0     7524
Name: count, dtype: int64

cancelou
0.0    73.0%
1.0    27.0%
Name: proportion, dtype: object

Com As Mudanças Aplicadas o percentual de cancelamentos caio para 27%
 é claro que 100% é utópico, mas com isso já temos as principais causas (ou talvez 3 das principais):
- forma de contrato mensal
- necessidade de ligações no call center
- atraso no pagamento